src/CoreWCF.Primitives/src/CoreWCF/IdentityModel/Tokens/Saml2SecurityToken.cs

```diff
@@ -0,0 +1,116 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.ObjectModel;
+using CoreWCF.Security;
+using MSIdentitySAML2 = Microsoft.IdentityModel.Tokens.Saml2;
+
+namespace CoreWCF.IdentityModel.Tokens
+{
+    /// <summary>
+    /// A security token backed by a SAML2 assertion.
+    /// </summary>
+    public class Saml2SecurityToken : SecurityToken
+    {
+        private MSIdentitySAML2.Saml2Assertion _assertion;
+        private readonly MSIdentitySAML2.Saml2SecurityToken _msIdentitySAML2SecurityToken;
+        private SecurityToken _signingToken;
+        private string _assertionXML;
+
+        protected Saml2SecurityToken()
+        {
+        }
+
+        public Saml2SecurityToken(MSIdentitySAML2.Saml2SecurityToken saml2SecurityToken)
+        {
+            _msIdentitySAML2SecurityToken = saml2SecurityToken;
+            _assertion = _msIdentitySAML2SecurityToken.Assertion;
+
+        }
+
+        public Saml2SecurityToken(MSIdentitySAML2.Saml2Assertion assertion)
+        {
+            Initialize(assertion);
+        }
+
+        protected void Initialize(MSIdentitySAML2.Saml2Assertion assertion)
+        {
+            if (assertion == null)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(assertion));
+
+            _assertion = assertion;
+            // this.assertion.MakeReadOnly();
+        }
+
+        public string AssertionXML
+        {
+            get { return _assertionXML; }
+            set { _assertionXML = value; }
+        }
+
+        public override string Id
+        {
+            get { return _assertion.Id.Value; }
+        }
+
+        //public override ReadOnlyCollection<SecurityKey> SecurityKeys
+        //{
+        //    get
+        //    {
+        //        return this.assertion.SecurityKeys;
+        //    }
+        //}
+
+        public MSIdentitySAML2.Saml2SecurityToken InternalSaml2SecurityToken
+        {
+            get { return _msIdentitySAML2SecurityToken; }
+        }
+
+        public MSIdentitySAML2.Saml2Assertion Assertion
+        {
+            get { return _assertion; }
+        }
+
+        public SecurityToken SigningToken
+        {
```

Comment:
This needs to be pulled from the IM Security Token. You should just be able to pull the Xml for the SigningToken and use the SecurityTokenHandlerCollectionManager to find the correct serializer to read the signing token.